File: progs/javascript/periodic_halt_check.js
// HaltedError class
class HaltedError extends Error {
  constructor() {
    super('Operation halted');
    this.name = 'HaltedError';
  }
}

// HaltSignal class (similar to Rust's Arc<AtomicBool>)
class HaltSignal {
  constructor() {
    this.halted = false;
  }

  halt() {
    this.halted = true;
  }

  isHalted() {
    return this.halted;
  }

  reset() {
    this.halted = false;
  }
}

function sleep(ms, signal) {
  return new Promise((resolve, reject) => {
    const timeoutId = setTimeout(resolve, ms);

    const abortHandler = () => {
      clearTimeout(timeoutId);
      reject(new Error('Sleep operation cancelled'));
    };

    signal?.addEventListener('abort', abortHandler);
  });
}

async function runWithPeriodicHaltCheck(workPromise, haltSignal, checkIntervalMs) {
  const controller = new AbortController();
  const signal = controller.signal;

  const periodicCheck = async () => {
    while (!signal.aborted) {
      await sleep(checkIntervalMs);
      if (haltSignal.isHalted()) {
        controller.abort();
        throw new HaltedError();
      }
    }
  };

  try {
    return await Promise.race([
      workPromise(signal),
      periodicCheck()
    ]);
  } finally {
    controller.abort();
  }
}

// Async task function (equivalent to my_async_task)
async function myAsyncTask(signal) {
  console.log("Task: Starting work...");
  await sleep(1000, signal);
  await sleep(1000, signal);
  await sleep(1000, signal);
  console.log("Task: Finished work.");
  return "Task Completed Successfully";
}

// Main execution function
async function main() {
  // Scenario 1: Running task, expecting completion
  console.log("Scenario 1: Running task, expecting completion.");
  const haltSignal1 = new HaltSignal();
  const checkInterval = 500; // 500ms

  try {
    const result1 = await runWithPeriodicHaltCheck(
      myAsyncTask,
      haltSignal1,
      checkInterval
    );
    console.log(`Scenario 1 Result: ${result1}`);
  } catch (error) {
    console.log(`Scenario 1 Error: ${error.message}`);
  }
  console.log("--------------------");

  // Scenario 2: Running task, expecting halt
  console.log("Scenario 2: Running task, expecting halt.");
  const haltSignal2 = new HaltSignal();

  // Start the task with halt checking
  const scenario2Promise = (async () => {
    try {
      const result2 = await runWithPeriodicHaltCheck(
        myAsyncTask,
        haltSignal2,
        checkInterval
      );
      console.log(`Scenario 2 Result: ${result2}`);
    } catch (error) {
      console.log(`Scenario 2 Error: ${error.message}`);
    }
  })();

  // Schedule halt signal after 7 seconds
  setTimeout(() => {
    console.log("Scenario 2: Setting halt signal!");
    haltSignal2.halt();
  }, 1000);

  await scenario2Promise;
  console.log("--------------------");
}

// Run the main function
main().catch(error => {
  console.error("Unhandled error:", error);
});

File: progs/python/misc/periodic_halt_check.py
import asyncio
import threading
from dataclasses import dataclass
from typing import TypeVar, Generic, Callable, Awaitable, cast


@dataclass
class HaltedError(Exception):
    def __str__(self) -> str:
        return "Operation halted"


T = TypeVar('T')

async def run_with_periodic_halt_check(
    work: Callable[[], Awaitable[T]],
    halt_signal: threading.Event,
    check_interval: float
) -> T:
    work_task = asyncio.create_task(work())
    while True:
        done, pending = await asyncio.wait([work_task], timeout=check_interval)
        if work_task in done:
            return done.pop().result()
        elif halt_signal.is_set():
            try:
                work_task.cancel()
                await work_task
            finally:
                raise HaltedError()


async def my_async_task() -> str:
    print("Task: Starting work...")
    await asyncio.sleep(1)
    await asyncio.sleep(1)
    await asyncio.sleep(1)
    print("Task: Finished work.")
    return "Task Completed Successfully"


async def main():
    """Run the example scenarios."""
    halt_signal = threading.Event()
    check_interval = 0.5  # 500ms
    
    print("Scenario 1: Running task, expecting completion.")
    try:
        result = await run_with_periodic_halt_check(
            my_async_task,
            halt_signal,
            check_interval
        )
        print(f"Scenario 1 Result: {result}")
    except HaltedError as e:
        print(f"Scenario 1 Error: {e}")
    
    print("--------------------")
    
    print("Scenario 2: Running task, expecting halt.")
    # Reset the halt signal
    halt_signal.clear()
    
    # Create a task to set the halt signal after 7 seconds
    async def set_halt_signal():
        await asyncio.sleep(1)
        print("Scenario 2: Setting halt signal!")
        halt_signal.set()
    
    # Start the task to set the halt signal
    asyncio.create_task(set_halt_signal())
    
    try:
        result = await run_with_periodic_halt_check(
            my_async_task,
            halt_signal,
            check_interval
        )
        print(f"Scenario 2 Result: {result}")
    except HaltedError as e:
        print(f"Scenario 2 Error: {e}")
    
    print("--------------------")
    await asyncio.sleep(2)


if __name__ == "__main__":
    asyncio.run(main())
